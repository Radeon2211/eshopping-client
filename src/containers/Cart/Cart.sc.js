import styled from 'styled-components';

export const PayBox = styled.div`
  align-items: center;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-bottom: ${({ theme }) => theme.spacings.level2};

  & .to-pay-text {
    font-size: ${({ theme }) => theme.fontSizes.level4};
  }

  & .to-pay-value {
    font-size: 2.6rem;
    letter-spacing: 1px;
    margin-left: ${({ theme }) => theme.spacings.level1};
  }
`;

export const Summary = styled.div`
  width: 100%;
  position: sticky;
  top: 9rem;

  @media only screen and (max-width: 56.25em) {
    bottom: -1px;
    padding-bottom: 1px;
  }

  & > *:first-child {
    transition: padding ${({ theme }) => theme.durations.level2}s;
  }

  &.is-sticky {
    border-top: 1px solid ${({ theme }) => theme.colors.light3};

    & > *:first-child {
      padding: ${({ theme }) => theme.spacings.level2};
    }
  }
`;
