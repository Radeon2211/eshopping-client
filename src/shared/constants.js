import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
