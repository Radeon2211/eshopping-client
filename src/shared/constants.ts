import * as Yup from 'yup';
import countryCodes from 'country-codes-list';
import { ProductCondition, ProductPhotoFieldValue } from './types/enums';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([ProductCondition.NEW, ProductCondition.USED, ProductCondition.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
  photo: Yup.mixed()
    .test('is-error', (value) => value !== ProductPhotoFieldValue.ERROR)
    .nullable(),
};

export const listOfAreaCodes = countryCodes.customArray({
  value: '{countryCallingCode}',
  label: '+{countryCallingCode} {countryNameEn}',
});

export const listOfCountries = countryCodes.customArray({
  value: '{countryNameEn}',
  label: '{countryNameEn}',
});

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: 'Price - ascending' },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: 'Price - ascending' },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig: ScrollToOptions = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};
