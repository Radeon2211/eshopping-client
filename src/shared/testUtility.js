// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
