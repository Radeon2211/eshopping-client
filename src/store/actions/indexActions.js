export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
