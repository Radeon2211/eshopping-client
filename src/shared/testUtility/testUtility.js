// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
