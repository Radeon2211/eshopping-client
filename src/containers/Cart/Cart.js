import React, { useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useHistory } from 'react-router-dom';
import * as SC from './Cart.sc';
import * as actions from '../../store/actions/indexActions';
import Heading from '../../components/UI/Heading/Heading';
import SideBySide from '../../components/UI/SideBySide';
import PlainPanel from '../../components/UI/Panels/PlainPanel';
import StickyPanel from '../../components/UI/Panels/StickyPanel';
import Button from '../../components/UI/Button/Button';
import Loader from '../../components/UI/Loader/Loader';
import LoadingOverlay from '../../components/UI/LoadingOverlay/LoadingOverlay';
import ToPayInfo from '../../components/UI/ToPayInfo/ToPayInfo';
import CartAndTransactionItems from '../../components/CartAndTransactionItems/CartAndTransactionItems';
import FlexWrapper from '../../components/UI/FlexWrapper';
import { itemTypes, defaultAppPath, modalTypes } from '../../shared/constants';
import { roundOverallPrice } from '../../shared/utility/utility';
import { GreenText } from '../../styled/components';
import { ReactComponent as EmptyCart } from '../../images/empty-cart.svg';

const Cart = () => {
  const history = useHistory();

  const isCartLoading = useSelector((state) => state.ui.isCartLoading);
  const cart = useSelector((state) => state.auth.cart);

  const dispatch = useDispatch();
  const onFetchCart = useCallback(() => dispatch(actions.fetchCart()), [dispatch]);
  const onGoToTransaction = useCallback(
    (currentHistory) => dispatch(actions.goToTransaction(currentHistory)),
    [dispatch],
  );
  const onSetModal = useCallback(
    (isModalOpen, modalContent) => dispatch(actions.setModal(isModalOpen, modalContent)),
    [dispatch],
  );

  useEffect(() => {
    onFetchCart();
  }, [onFetchCart]);

  let content = <Loader align="center" />;
  if (cart === null) {
    content = (
      <Heading variant="h3" mgBottom="h3" align="center" lineHeight="4">
        There is a problem to get your shopping cart
      </Heading>
    );
  } else if (cart) {
    if (cart.length > 0) {
      const cartValue = cart.reduce((acc, { product, quantity }) => {
        return acc + product.price * quantity;
      }, 0);
      const roundedCartValue = roundOverallPrice(cartValue);

      content = (
        <>
          <Heading variant="h3">Your shopping cart</Heading>
          <SideBySide proportion="3/1" makeVerticalWhen={1200}>
            <PlainPanel>
              <FlexWrapper justify="center">
                <Button color="red" clicked={() => onSetModal(true, modalTypes.CLEAR_CART)}>
                  clear the cart
                </Button>
              </FlexWrapper>
              <CartAndTransactionItems
                items={cart}
                type={itemTypes.CART}
                isCartLoading={isCartLoading}
              />
            </PlainPanel>
            <StickyPanel>
              <FlexWrapper direction="column" spacing="2">
                <ToPayInfo value={roundedCartValue} />
                <Button
                  filled
                  stretch
                  disabled={isCartLoading}
                  clicked={() => onGoToTransaction(history)}
                >
                  go to summary
                </Button>
              </FlexWrapper>
              {isCartLoading && <LoadingOverlay />}
            </StickyPanel>
          </SideBySide>
        </>
      );
    } else {
      content = (
        <SC.EmptyCart>
          <Heading variant="h3" mgBottom="3" align="center" lineHeight="4">
            Your shopping cart is empty. Check out the latest&nbsp;
            <Link to={defaultAppPath} data-testid="Cart-default-path-link">
              <GreenText>offers</GreenText>
            </Link>
          </Heading>
          <Heading variant="h4" align="center" lineHeight="4" mgBottom="3">
            You can have up to 50 products in the cart
          </Heading>
          <EmptyCart className="empty-cart-image" />
        </SC.EmptyCart>
      );
    }
  }

  return content;
};

export default Cart;
