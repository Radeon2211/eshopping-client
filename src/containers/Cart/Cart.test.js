import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import thunk from 'redux-thunk';
import Cart from './Cart';
import theme from '../../styled/theme';
import { createCartItem } from '../../shared/testUtility/testUtility';
import { defaultAppPath, modalTypes } from '../../shared/constants';
import * as actions from '../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const defaultCartItem = createCartItem({
  sellerUsername: 'user1',
  quantity: 5,
  productId: 'p1',
  price: 499.97,
});

const setUp = (cart, isCartLoading = false, pushFn = jest.fn()) => {
  const history = {
    listen: jest.fn(),
    createHref: jest.fn(),
    location: { pathname: '/cart' },
    push: pushFn,
  };

  const store = mockStore({
    auth: { cart },
    ui: { isCartLoading },
  });
  store.dispatch = jest.fn();

  return {
    ...render(
      <Router history={history}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <Cart />
          </ThemeProvider>
        </Provider>
      </Router>,
    ),
    store,
    history,
  };
};

jest.mock('../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../store/actions/indexActions.js'),
  fetchCart: () => {},
  goToTransaction: (history) => history,
}));

afterEach(cleanup);

describe('<Cart />', () => {
  describe('Check how renders', () => {
    it('should render everything correctly with one cart item', () => {
      const { asFragment } = setUp([defaultCartItem]);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render empty cart', () => {
      const { asFragment } = setUp([]);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render only <Loader />', () => {
      const { asFragment } = setUp(undefined, true);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render that there is a problem to fetch cart', () => {
      const { asFragment } = setUp(null);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render <LoadingOverlay /> and go to summary button should be disabled', () => {
      setUp(
        [
          createCartItem({
            sellerUsername: 'user1',
            quantity: 5,
            productId: 'p1',
            price: 499.97,
          }),
        ],
        true,
      );
      expect(screen.getByTestId('LoadingOverlay')).toBeInTheDocument();
      expect(screen.getByText('go to summary')).toBeDisabled();
    });
  });

  describe('Check behaviour after clicking and calling redux actions', () => {
    it('should call push with defaultAppPath after click on link in empty cart', () => {
      const pushFn = jest.fn();
      setUp([], false, pushFn);
      fireEvent.click(screen.getByTestId('Cart-default-path-link'));
      expect(pushFn).toHaveBeenCalledTimes(1);
      expect(pushFn).toHaveBeenCalledWith(defaultAppPath);
    });

    it('should call fetchCart() and also goToTransaction() and setModal()', () => {
      const { store, history } = setUp([defaultCartItem]);

      expect(store.dispatch).toHaveBeenNthCalledWith(1, actions.fetchCart());

      fireEvent.click(screen.getByText('clear the cart'));
      expect(store.dispatch).toHaveBeenNthCalledWith(
        2,
        actions.setModal(true, modalTypes.CLEAR_CART),
      );

      fireEvent.click(screen.getByText('go to summary'));
      expect(store.dispatch).toHaveBeenNthCalledWith(3, actions.goToTransaction(history));
      expect(store.dispatch).toHaveBeenCalledTimes(3);
    });
  });
});
