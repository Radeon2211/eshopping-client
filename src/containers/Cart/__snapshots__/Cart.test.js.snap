// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> Check how renders should render empty cart 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ kpYzKn"
  >
    <h3
      class="sc-gsTCUz ezAATf"
    >
      Your shopping cart is empty. Check out the latestÂ 
      <a
        data-testid="Cart-default-path-link"
      >
        <span
          class="sc-fFubgz cjsnwZ"
        >
          offers
        </span>
      </a>
    </h3>
    <h4
      class="sc-gsTCUz cuzwDI"
    >
      You can have up to 50 products in the cart
    </h4>
    <svg
      class="empty-cart-image"
    >
      empty-cart.svg
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<Cart /> Check how renders should render everything correctly with one cart item 1`] = `
<DocumentFragment>
  <h3
    class="sc-gsTCUz iHHWjZ"
  >
    Your shopping cart
  </h3>
  <div
    class="sc-dlfnbm cQBmyc"
  >
    <section
      class="sc-hKgILt jLFCGR"
    >
      <div
        class="sc-jrAGrp fAfrlP"
        direction="row"
        spacing=""
        width="100%"
        wrap="nowrap"
      >
        <button
          class="sc-jSgupP icHxze"
          color="red"
          type="button"
        >
          clear the cart
        </button>
      </div>
      <div
        class="sc-idOhPF ikxypi"
      >
        <span
          class="sc-fubCfw jalAIu"
          color=""
          display="block"
          overflow="visible"
          size="3"
          spacing="0px"
          transform=""
        >
          <span>
            seller 
          </span>
          <a
            data-testid="CartAndTransactionItems-seller-link"
          >
            <span
              class="sc-fFubgz cjsnwZ"
            >
              user1
            </span>
          </a>
        </span>
        <div
          class="sc-bqyKva JQCJG"
          data-testid="CartItem"
        >
          <div
            class="photo-and-name"
          >
            <a
              data-testid="CartItem-product-link-photo"
            >
              <div
                class="sc-jrAGrp bGCJbw sc-dQppl bEkWdt"
                data-testid="ProductThumbnail"
                direction="row"
                height="7"
                spacing=""
                width="7"
                wrap="nowrap"
              >
                <img
                  alt="product name"
                  data-testid="ProductThumbnail-img"
                  src="no-photo.png"
                />
              </div>
            </a>
            <span
              class="name"
            >
              <a
                data-testid="CartItem-product-link-name"
              >
                product name
              </a>
            </span>
          </div>
          <div
            class="mobile-lower-row"
          >
            <div
              class="choose-quantity-box"
            >
              <div
                class="sc-fodVxV dsjgGH"
              >
                <button
                  class="button minus"
                  data-testid="ChooseQuantity-minus-btn"
                  type="button"
                >
                  <div
                    class="sc-hBEYos ejZCVz"
                    color=""
                    data-testid="ChooseQuantity-minus-icon"
                  >
                    <svg>
                      minus.svg
                    </svg>
                  </div>
                </button>
                <input
                  class="sc-kstrdz ceqwqy"
                  data-testid="NumberInput-quantity"
                  name="quantity"
                  type="number"
                  value="5"
                />
                <button
                  class="button plus"
                  data-testid="ChooseQuantity-plus-btn"
                  disabled=""
                  type="button"
                >
                  <div
                    class="sc-hBEYos dKYZQj"
                    color="#DADFE1"
                    data-testid="ChooseQuantity-plus-icon"
                  >
                    <svg>
                      plus.svg
                    </svg>
                  </div>
                </button>
              </div>
              <span
                class="sc-fubCfw bfbrJj"
                color=""
                data-testid="CartItem-available-quantity"
                display="inline-block"
                overflow="visible"
                size="2"
                spacing="0px"
                transform=""
              >
                of 2
              </span>
            </div>
            <div
              class="price-box"
            >
              <span
                class="sc-fubCfw ekxLGm"
                color=""
                display="inline-block"
                overflow="visible"
                size="5"
                spacing="0px"
                transform=""
              >
                $2,499.85
              </span>
              <span
                class="sc-fubCfw mDKSv"
                color=""
                data-testid="CartItem-price-per-piece"
                display="inline-block"
                overflow="visible"
                size="2"
                spacing="0px"
                transform=""
              >
                per piece $499.97
              </span>
            </div>
          </div>
          <div
            class="sc-hBEYos hrBfGW remove-icon"
            color="#F22613"
            data-testid="CartItem-trash-icon"
          >
            <svg>
              trash.svg
            </svg>
          </div>
        </div>
      </div>
    </section>
    <section
      class="sc-eCssSg bNvwyr"
    >
      <section
        class="sc-hKgILt jLFCGR"
      >
        <div
          class="sc-jrAGrp kdUhfw"
          direction="column"
          spacing="2"
          width="100%"
          wrap="nowrap"
        >
          <div
            class="sc-jrAGrp jvClWK"
            direction="row"
            spacing="1"
            width="100%"
            wrap="wrap"
          >
            <span
              class="sc-fubCfw bvkSGA"
              color=""
              display="inline-block"
              overflow="visible"
              size="4"
              spacing="0px"
              transform=""
            >
              To pay
            </span>
            <span
              class="sc-fubCfw cQoALv"
              color=""
              display="inline-block"
              overflow="visible"
              size="6"
              spacing="1px"
              transform=""
            >
              $2,499.85
            </span>
          </div>
          <button
            class="sc-jSgupP fYrlGR"
            color="blue"
            type="button"
          >
            go to summary
          </button>
        </div>
      </section>
    </section>
  </div>
</DocumentFragment>
`;

exports[`<Cart /> Check how renders should render only <Loader /> 1`] = `
<DocumentFragment>
  <div
    class="sc-gKsewC QQywS"
    data-testid="LoaderWrapper"
  >
    <div
      class="sc-iBPRYJ iftXuC"
      data-testid="Loader"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Cart /> Check how renders should render that there is a problem to fetch cart 1`] = `
<DocumentFragment>
  <h3
    class="sc-gsTCUz gYothf"
  >
    There is a problem to get your shopping cart
  </h3>
</DocumentFragment>
`;
